wordpress-mobile/WordPress-iOS
# Refactor MediaImageService and fullscreen image view

This PR refactors `MediaImageService` gets us 90% to the point where we’ll be able to remove the entire `MediaThumbnailCoordinator` and `MediaThumbnailService` that were replaced by a significantly improved async/await-based `MediaImageService` in 23.7.

Changes:

- Refactor `MediaImageService`:
    - Make it an actor and more more work to the background
    - Introduce private `SafeMedia` class, a thread-safe representation of `Media` and lift the requirements for most of the `MediaImageService` methods to be called from the main thread
    - It works better with the concurrency system and its requirements
- Move thumbnail memory cache to `MediaImageService`
- Add `MediaImageService/image(for:)` method to load original image
- Add MediaImageService-specific errors to replace URLError(.unknown) make debugging easier
- Update `WPImageViewController` to use the new API to fetch original images
- Move memory cache from `SiteMediaCollectionCellViewModel` to `MediaImageService`, which means the the memory cache now works for the page view controller in the Site Media screen that displays medium-sized thumbnails

To test:

- Verify that GIFs are rendered in fullscreen preview
- Verify that disk cache works (original images are loaded offline)
- Smoke test the Site Media screen thumbnails grid

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/ccf3da31-b721-4a2a-ade7-e97935ade181


## Regression Notes
1. Potential unintended areas of impact: Site Media
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual and unit tests
3. What automated tests I added (or what prevented me from doing so): yes

PR submission checklist:


UI Changes testing checklist:
